ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java

```diff
@@ -1030,6 +1037,315 @@ private Pair<ServerErrorCode, GetResponse> getBlobFromRemoteReplica(ReplicateBlo
+    if (replicateBlobRequest.getOperationType() == RequestOrResponseType.TtlUpdateRequest
+        || replicateBlobRequest.getOperationType() == RequestOrResponseType.DeleteRequest) {
+      // For RequestOrResponseType.PutRequest, we don't care about the life version.
+      if (replicateBlobRequest.getLifeVersion() != MessageInfo.LIFE_VERSION_FROM_FRONTEND) {
+        logger.info("ReplicateBlobRequest invalid life version {}", replicateBlobRequest);
+        completeReplicateRequest(request, replicateBlobRequest, ServerErrorCode.Bad_Request, startProcessTime);
+        return;
+      }
+    }
+
+    // this replica itself is the source replica. Don't need do anything. Return success status.
+    DataNodeId remoteDataNode = clusterMap.getDataNodeId(remoteHostName, remoteHostPort);
+    if (remoteDataNode != null && remoteDataNode.equals(currentNode)) {
+      logger.info("ReplicateBlobRequest this is the source replica, return immediately. {}", replicateBlobRequest);
+      completeReplicateRequest(request, replicateBlobRequest, ServerErrorCode.No_Error, startProcessTime);
+      return;
+    }
+
+    // check if local store has the key already
+    ServerErrorCode errorCode = ServerErrorCode.No_Error;
+    MessageInfo localMessageInfo = null;
+    try {
+      StoreKey convertedKey = getConvertedStoreKeys(Collections.singletonList(blobId)).get(0);
+      if (!(convertedKey).equals(blobId)) {
+        logger.info("ReplicateBlobRequest replicated Blob {} convert to local key {}", blobId, (BlobId) convertedKey);
+        blobId = (BlobId) convertedKey;
+      }
+      Store store = storeManager.getStore(blobId.getPartition());
+      localMessageInfo = store.findKey(convertedKey);
+      logger.info("ReplicateBlobRequest replicated Blob {}, local Store has the Key", blobId);
+    } catch (StoreException e) {
+      // store.findKey throws e.getErrorCode() == StoreErrorCodes.ID_Not_Found if it doesn't exist.
+      if (e.getErrorCode() == StoreErrorCodes.ID_Not_Found) {
+        errorCode = ServerErrorCode.No_Error;
+      } else {
+        logger.error("ReplicateBlobRequest store.findKey throw exception {} on Blob {},", e.getErrorCode(), blobId);
+        errorCode = ErrorMapping.getStoreErrorMapping(e.getErrorCode());
+      }
+    } catch (Exception e) {
+      // getConvertedStoreKeys which may throw Exception
+      logger.error("ReplicateBlobRequest getConvertedStoreKeys unknown exception to replicate {} of {}", blobId,
+          replicateBlobRequest, e);
+      errorCode = ServerErrorCode.Unknown_Error;
+    }
+    if (errorCode != ServerErrorCode.No_Error || (
+        replicateBlobRequest.getOperationType() == RequestOrResponseType.PutRequest && localMessageInfo != null)) {
+      // If we want to replicate PutBlob and local host has the key already, return.
+      completeReplicateRequest(request, replicateBlobRequest, errorCode, startProcessTime);
+      return;
+    }
```

Comment:
I don't think we ever want to repair a put operation.

Reply:
As above. It supports V1.